Expansion and Improvements to Classical/Biblical Hebrew

EricJHarvey opened the pull request:
This pull request expands and improves the Classical/Biblical Hebrew tables (Israeli Hebrew tables remain unchanged).
Two new translation tables have been added so that the three main standards for Classical Hebrew are now available:
1. The original International Hebrew Braille Code (hbo.utb)
2. The system for full cantillation developed by Ray McAllister, Matthew Yeater, and Sarah Blake LaRose (hbo-cantillated.utb)
3. The compact standard without dagesh or shewa represented in the Katz transcription manual (hbo-slim.utb)
Full references for documentation are provided in the metadata for each file. I will also submit a braille spec soon.

I made hbo.utb the table for the International Hebrew Braille Code. In v. 3.33, hbo.utb is used for the IHBC-McCallister system (hbo-cantillated.utb in this PR). The International Hebrew Braille Code is the most-established and widely used standard for Classical Hebrew, so it seems right to make it the basic table, but if this causes problems with reverse-compatibility I can change it. 

A few other notes on the tables:
1. Most of the Hebrew-specific rules are contained in .uti files, and the.utb files only add rules for uncontracted English support. Since Classical Hebrew readers usually have another first language, this allows non-English speakers to create custom translation tables with Hebrew and their first language, or for easier inclusion in multilingual tables like ancient-languages-us.utb..

In addition to the new tables, this PR makes improvements to the hbo-cantilated.utb table:
In general, I switched to a rules-based approach so the tables could share the common character files and use far fewer lines of code overall. 

I also made some improvements to hbo-cantillated.utb for consistency and to make sure that braille output matches that produced by Duxbury as closely as possible. Specifics follow.
- Ultralong vowels now contract consistently.
- Cantillation marks no longer interrupt the contractions for ultra-long vowels (hiriq-yod and tsere-yod).
- Holam now always renders properly as 135 and holam waw as 246.

Because of this, I had to change some of the tests in the yaml file. I know that yaml tests ideally should not change, so I will provide details of all the line changes in the next comment.

EricJHarvey commented:
@yeatersink @paulGeoghegan @leonarddeR @sjlarose Please review!

LeonarddeR commented:
First of all, I really want to congratulate @EricJHarvey with this stunning result. I'm very glad to see a set of tables that fully utilize the power of liblouis (correct/context/pass opcodes, virtual dots, etc.). I am impressed by the compactness and quality of these tables.
I did a run down of everything and created [a small pull request](https://github.com/EricJHarvey/liblouis/pull/2) to improve consistency of metadata and usage of comments. With that I hope to prevent polluting the current Pull Request with a lot of inline commentary.

EricJHarvey commented:
Thank you, @LeonarddeR! Those fixes were very helpful.

EricJHarvey commented:
Awesome! Thanks, Matt.

LeonarddeR commented:
Just created https://github.com/EricJHarvey/liblouis/pull/3, that should get the ucs4 build working with passing tests.
Only thing remaining then is regenerating the display-names file. That process definitely needs documentation.

egli commented on tables/hbo-cantillated-rules.uti:
```diff
@@ -0,0 +1,70 @@
+# developed by Eric J. Harvey with support from the national
+# Endowment for the Humanities (USA) and Stanford University.
+# Many thanks to Leonard de Ruijter and Sarah Blake LaRose for their
+# help developing and testing the tables, and to Abbie Howell for
```
I think her name is Abby Howell ('y' instead of 'ie')

egli commented on tables/hbo-cantillated.utb:
```diff
@@ -0,0 +1,109 @@
+# developed by Eric J. Harvey with support from the national
+# Endowment for the Humanities (USA) and Stanford University.
+# Many thanks to Leonard de Ruijter and Sarah Blake LaRose for their
+# help developing and testing the tables, and to Abbie Howell for
```
Again, her name is Abby

egli commented on tables/hbo-slim.utb:
```diff
@@ -0,0 +1,100 @@
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+# liblouis  comes with ABSOLUTELY NO WARRANTY.
+
+# The first Hebrew Braille table hosted on Lib Louis was developed by
```
I sound like a nag, but the official terminology is "liblouis", not "Lib Louis" :-)

egli reviewed (changes requested):
Great work! Thanks Eric. If we can get the tests to pass then I wholeheartedly approve

bertfrees commented:
> I was unable to specify tables using language and __assert-match: because it always preferred hbo-slim.uti

I'll try to fix that.

bertfrees commented on tests/braille-specs/hbo.yaml:
```diff
@@ -140,18 +340,3 @@ tests:
   - ["והארץ", "⠺⠓⠁⠗⠮"]
   - ["המים", "⠓⠍⠚⠍"]
-
-table: he-common-vowels-ihbc.uti
```
@EricJHarvey You said:

> I removed the tests for he-comon-vowels-ihbc.uti, since the refactoring means this table no longer functions on its own.

Could it perhaps be replaced with a sequence of table files?

You did not only remove tests for he-comon-vowels-ihbc.uti, but you also removed (the same) tests for hbo.utb, ancient-languages-borger.utb and ancient-languages-us.utb. This is because tests are run against all previously declared tables. Could we add these back, or are those test cases covered elsewhere now?

bertfrees replied:
I checked that all of these six tests except the first one still work if you apply them to hbo.utb or hbo-cantillated.utb.

bertfrees replied:
I was able to put most tests that were deleted back, except that one I mentioned:

- ["בְּרֵאשִׁית", "⠃⠄⠗⠌⠁⠩⠊⠚⠹"]

and also there seems to be a problem with the ancient-languages-borger.utb and ancient-languages-us.utb tests. Something has possibly gone wrong there.

EricJHarvey replied:
> I was able to put most tests that were deleted back, except that one I mentioned:
> 
> * ["בְּרֵאשִׁית", "⠃⠄⠗⠌⠁⠩⠊⠚⠹"]
> 
> and also there seems to be a problem with the ancient-languages-borger.utb and ancient-languages-us.utb tests. Something has possibly gone wrong there.

@bertfrees Thanks for finding this solution. The failing test was another case of hiriq-yod being shown uncontracted where it should have been contracted. I have fixed it and I get passing tests now.

EricJHarvey commented:
Thanks all. I can look over these and merge Leonard's PR in about an hour.Eric J. Harvey, Ph.D.He/hisAffiliated ResearcherCenter for Spatial and Textual AnalysisStanford Universitywww.blindscholar.comOn May 28, 2025, at 6:53 AM, Bert Frees ***@***.***> wrote:﻿
@bertfrees commented on this pull request.



In tests/braille-specs/hbo.yaml:
> @@ -140,18 +340,3 @@ tests:
   - ["היתה", "⠓⠚⠹⠓"]
   - ["והארץ", "⠺⠓⠁⠗⠮"]
   - ["המים", "⠓⠍⠚⠍"]
-
-table: he-common-vowels-ihbc.uti

Oh cool. Missed that. I'll check it out. Thanks!

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

EricJHarvey reviewed (commented):
> Again, her name is Abby 

Thanks for the edits! Not taken as nagging at all. I've fixed them where they appear.

EricJHarvey reviewed (commented):
> Again, her name is Abby 

Thanks for the edits! Not taken as nagging at all. I've fixed them where they appear.

EricJHarvey commented:
@egli @bertfrees @LeonarddeR I think I've resolved all of the issues I can. Getting the final check to pass is beyond my abilities. Thanks!

bertfrees commented on tables/hbo-cantillated.utb:
```diff
@@ -0,0 +1,109 @@
+#+language: xdm
+#+language: phn
+#+type: literary
+#+contraction: yes
```
The `contraction` field normally has the values `no`, `full` or `partial`. Which of these matches best?

EricJHarvey commented:
Yes, I would agree with "partial." The only contractions are two cells contracting into a single cell. There are no word- or syllable-level contractions.On May 29, 2025, at 5:41 AM, Leonard de Ruijter ***@***.***> wrote:﻿
@LeonarddeR commented on this pull request.



In tables/hbo-cantillated.utb:
> @@ -0,0 +1,109 @@
+# liblouis: Classical Hebrew, Fully cantillated, IHBC-McAllister
+#
+#-index-name: Hebrew, classical, cantillated
+#-display-name: Classical Hebrew braille with full cantillation
+#
+#+language: hbo
+#+language: oar
+#+language: arc
+#+language: jpa
+#+language: obm
+#+language: xdm
+#+language: phn
+#+type: literary
+#+contraction: yes

Partial makes the most sense here.
Should file extentions actually become CTB?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

EricJHarvey commented:
> I'm not sure if [EricJHarvey#3](https://github.com/EricJHarvey/liblouis/pull/3) actually did the trick. I'm trying to enable the ancient-languages-x tests again, but I'm getting a lot of failures. Don't know what I'm doing wrong.

Yes, I'm now getting 46 failures. It looks like some tests are being tested for back-translation when they shouldn't. I'll go through this and see what I can do.

yeatersink commented:
Hey everyone, 
After looking at some things and consulting with a couple people who use this table, I think that it is best to interject. 
The key problem with this table is that it does not take into consideration accents that are pre and post positive. They can be both.
 The best way to deal with this is to hard code every possibility into the table, so lib louis gets it right every time. 

Not trying to be negative, but this  table seems to have the same problems as our original table. There are 2 key problems. 1 the Hebrew and 2 the how the machine reads the Unicode. I will present the problems below to help address the problems. 
For clarity I will use terminology in the following way. 1 character refers to the 4 stanzas of Unicode.  2 Stanza, is a tiny bit of Unicode which make up 4 parts of a Unicode character. 

The first Problem:
1 The Hebrew. 
Hebrew is written with a 4 stanza Unicode. 1 Consonant. 2 Vowel. 3 Cantillation mark. The other is the dagesh. I did not number it because it causes the most problems. Nevertheless, that is where all 4 stanzas come from in the Hebrew character. 

The other thing that causes the problems are the cantillation marks, and weather they are pre or post positive.  That is to say if they come before or after the vowel… This is why the heriq yod is not always contracted. If you do not know if the accent is intended to be pre or post positive, then it will create a problem because the heriq yod should be contracted and some times not. In fact this creates many problems with other vowels because of the ordering of the stanzas. If this is not correct, then it will thro the incorrect braille. 
Just because we get a test that passes, does not mean that problems are not created in other areas. This is what we are seeing now with this table.  

Second problem:
2 The way the Machine reads the Unicode:
The problem is that the machine needs to read all 4 stanzas and then assign the correct braille to that Unicode character. This should not be a problem but it is. The reason is because if it reads a character, with all stanzas, and you do not have a table with that exact character, with that exact ordering of stanzas, then it will assign what it thinks should be there or throw an error. This is why it produces the incorrect braille. 

The most frustrating part is that just when you think that you caught the fox that was creating the wrong braille, it creates problems in other areas, unless you have the correct braille in the table for the character  the machine thinks that it is reading. 

The other major issue is the person creating the document that is being brailled. There are all kinds of reasons that the stanzas of a Hebrew character get disorganized, and this must be taken into consideration. 

So what is the solution?
The solution is to have a table that takes all of this into consideration. It is not easy, but it is the only effective way of providing accurate braille in every context. Which is why I provided the test/ yaml document that I did. I took material from 5 major Hebrew literature sources to ensure that the table was   going to produce the correct braille in all of those contexts.

 This is why Paul and I created and published and released Niv Louie. Niv Louie makes dealing with these problems easy, and super fast.

@egli  @bertfrees 
This table that Eric is attempting to merge does not take into consideration the problems that plagued us in the beginning. This is what is causing the problems now in these tests failing. 
I am afraid that this table is a step backward and will create many more problems than it attempts to repair. It will do so in many areas. 

My suggestion is to include every possibility the stanza could be written, so Lib Louis does not have the ability to make a mistake. This includes pre and post positive accents. 
Every plausible stanza must be hard coded into the table so Lib Louis produces the correct braille in every context. 
To do this by hand would take ages. This is why Paul and I released Niv Louie, it makes it easy and fast. And it creates test/ yaml documents so you can test vasts pieces of literature very quickly and efficiently.

EricJHarvey commented:
Hi @yeatersink 
I’m open to the idea that these tables aren’t perfect and may still need tweaking, but we can’t speak in generalities.

We can’t talk about what braille is “correct” without reference to a specific standard. These tables translate according to three different standards, which do not treat the vowels and cantillation the same. Therefore, the tests in the yaml file are divided into sections pertaining to each table.
Any issue with prepositive or postpositive cantillation only applies to the IHBC-McAllister system, which is translated here according to the documentation on the Duxbury site. 

I have given reference information for each standard in the table metadata. All previous yaml tests have been retained, and I have added more to cover the two additional standards.
If you would like to dispute translations, please provide specific information.

This could come in the form of line comments in the yaml file. I provided a list of changes in my second comment above. You could also review the full yaml file for issues. There are now many more tests than there were before. For this, please provide:
1. The line that produces incorrect braille.
2. Proposed correction.
3. A reason why the braille is incorrect, referencing relevant documentation.

Alternatively, you could provide a new text or file that does not translate correctly, with documentation of the reason.

In general, I think a rules-based approach leads to more accurate and consistent translation than hard-coding all possibilities, but that is something that can only be proven with specific cases.

LeonarddeR commented:
Hi @yeatersink

Pre and post positive accents are new to me. I am happy to have learned something new.

That said, you provide a lot of detail, but unfortunately without any examples. It would help to have some examples that:
1. Succeed in liblouis 3.33 (your table)
2. Fail in the table in this current table (erics table, notably hbo-cantillated.utb).

I agree with @EricJHarvey that a rules based approach is definitely the way to go here. That's how all more complex braille codes are handled in liblouis anyway.

yeatersink commented:
Certainly! Here's a concrete example from a biblical verse that illustrates how cantillation marks can appear before the vowel or syllable they govern in chant notation.
Example from Psalm 23:1 (Tehillim 23:1):

23 1‮מִזְמֹ֥ור לְדָוִ֑ד יְהוָ֥ה רֹ֝עִ֗י לֹ֣א אֶחְסָֽר׃

2 Words: מִזְמֹ֥ור   and רֹ֝עִ֗י

You can see that the cantillation mark is placed in the middle of the holam and vav. And in the second word, between the heriq and yud. 
The markings appear visually either above or below a letter, but the Unicode places in the middle of these letters because of how they are written. 
In cantillation literature which is indeed classical literature, the cantillation marks are placed intentionally in front of the vowel to underscore pronunciation especially in contexts where  the text is meant to be chanted, which includes the Hebrew bible as well as other literature.  This needs to be taken into consideration. 

Thus, if you have holam vuv coded with out the cantillation mark, this will throw an error or produce wrong braille.  The same for the heriq yud. This is one example of many that are available.   

There are times when the holum vav and the heriq yud appear with out cantillation marks. So there may not be a problem there.  

I am very happy that you are working on this. I have like three jobs and 2 grandbabies so I don’t have much time to work on everything; especially since there are very capable people like you to do it. All I  am suggesting is that you include these issues so that there are not errors in the areas that tests being submitted do not cover. Or there will inevitably be another pull request.

LeonarddeR commented:
Thanks for your example. Could you please provide the unicode braille you'd expect for these words? That can help in finding out what you'd expect from the table.
According what @EricJHarvey mentioned in https://github.com/liblouis/liblouis/pull/1787#issuecomment-2919975645, the following would help with required to both מִזְמֹ֥ור and רֹ֝עִ֗י:

2. Proposed correction (in unicode braille)
3. A reason why the braille is incorrect, referencing relevant documentation. So specifically, it is not strictly relevant here how things look visually, it is relevant how it should be brailled according to the documentation.

EricJHarvey commented:
@yeatersink 
That’s a super interesting edge case. Where did you find the text? I’d like to look at the source and see if they do any other strange things with the encoding.
I ran the text through the new tables and got this:
⠼⠃⠉ ⠼⠁ ⠍⠊⠵⠄⠍⠕⠨⠍⠺⠗ ⠇⠄⠙⠣⠺⠊⠆⠙ ⠚⠄⠓⠺⠣⠨⠍⠓ ⠗⠕⠨⠡⠫⠨⠗⠔ ⠇⠕⠨⠥⠁ ⠁⠑⠭⠄⠎⠣⠈⠗⠲
The new tables already translate רֹ֝עִ֗י properly, but the מִזְמֹ֥ור does not produce proper braille. Technically, it is not correct unicode either, since the holam should be attached as a combining character to the waw, not the mem, but the visual characters display in such a way that it can look right if the reader is not looking closely, so it could appear sometimes out in the wild.
I looked for texts online and in Accordance that encoded the holam waw with the holam before the waw this way, but didn’t find any. So please do let me know where it came from.

I wrote a rule to account for it in the translation table and added this text to the yaml file as a test.
Now the table gives this as output for the same text:
⠼⠃⠉ ⠼⠁⠍⠊⠵⠄⠍⠨⠍⠪⠗ ⠇⠄⠙⠣⠺⠊⠆⠙ ⠚⠄⠓⠺⠣⠨⠍⠓ ⠗⠕⠨⠡⠫⠨⠗⠔ ⠇⠕⠨⠥⠁ ⠁⠑⠭⠄⠎⠣⠈⠗⠲

Which is identical to the way it renders other character orders of the same text, such as  this from Accordance: Psa. 23:1 מִזְמ֥וֹר לְדָוִ֑ד יְהוָ֥ה רֹ֝עִ֗י לֹ֣א אֶחְסָֽר׃
⠠⠏⠎⠁⠲ ⠼⠃⠉⠒⠼⠁ ⠍⠊⠵⠄⠍⠨⠍⠪⠗ ⠇⠄⠙⠣⠺⠊⠆⠙ ⠚⠄⠓⠺⠣⠨⠍⠓ ⠗⠕⠨⠡⠫⠨⠗⠔ ⠇⠕⠨⠥⠁ ⠁⠑⠭⠄⠎⠣⠈⠗⠲

This text also clued me in to the existence of the “right-to-left override” characters, which I did not realize were distinct from the “left-to-right mark” characters. I’ve added rules to remove them so the hex code doesn’t clutter the braille.

Let me know what you think of the braille here and if you identify any other potential issues.

LeonarddeR commented:
@EricJHarvey your rule now only covers ֹו in cases there's an accent between the holam and the waw. I wonder whether there are cases where holam and waw are directly placed in sequence where it is supposed to be a holam waw vowel.
The new rule might also contract holam and waw incorrectly when the waw is a consonant.

LeonarddeR commented:
Great @EricJHarvey, that's exactly what I aimed at.
should the rule check you added yesterday for accents also be extended in such a way that it will only apply when the waw is directly followed by another consonant?

EricJHarvey commented:
@LeonarddeR 
I’ve been trying to make this work, but unfortunately I don’t think it’s possible with the context and multipass opcode syntax constraints. We would need to be able to have two sets of replacement brackets in the test part, which is not supported.
I saw what you were intending with the slash before %hebLetter, but couldn’t get it to work in practice.
Fortunately, I don’t think it’s necessary for this rule.
We already have specified correct behavior for the character sequence <holam><waw><vowel>, so we are only concerned with the sequence <holam><Accent><waw><vowel>. This should never occur.
TO begin with, a holam followed by a consonantal waw is extremely ungrammatical in biblical Hebrew.
It only occurs regularly in one context: the perpetual qere of the divine name. This is the exception that proves the rule, because it is ungrammatical on purpose. It is a “ghost word” created by combining the consonants with vowels from a different word, Adonai. And since cantillation marks only occur on the yod, waw, or second he, the holam will never be followed by an accent. So it will always braille properly:
יְהֹוָ֣ה
⠚⠄⠓⠕⠺⠣⠨⠥⠓
I spent a while searching for other situations in which a holam, cantillation mark, waw might be followed by a vowel and came up empty. 
In other words, I’d be happy to constrain this rule to contexts where the waw is followed directly by a consonant, but a) I don’t know how, and b) I don’t think it’s necessary.
Let me know if you find a way.

egli reviewed (approved):
Very solid work. Thank you very much

EricJHarvey commented:
@bertfrees And in regard to the question of regression, the short answer is no. The two words Matt brought up both translated incorrectly in the 3.33 tables. One was already fixed by this PR and the other one now also translates correctly. The open issue relates to a broader question raised by that word.

yeatersink commented:
Hey all, 

I think that there is some misunderstanding about the BHS and the Cantillation marks in general.

The BHS is the most widely used version of the Hebrew Bible and the facts are that there are several versions of the BHS. The other major issue is that the BHS predominantly engages with the Masoretic Text (MT) and the   facts are that the MT varies based upon the cantillation marks depending on what tradition the MT is representing. 

The comments by Eric demonstrate a lack of understanding of these critical issues. I am not saying this to stop this table from being published, but I am making these comments because of the comments of what Eric thinks are “not best practices.” 

The reality is that there is much in lack of understanding in these issues and there is still much work to do on this table. 

The facts are that the cantillation marks can go above or below consonants   and they can go above or below vowels.  Depending on the tradition, they can be placed in the Unicode in what ever sequence is needed to place the cantillation mark in the space needed to reflect the desired pronunciation.  

In my initial comments on Psalms 23, I literally found that example of where the tests in this table did not cover, to show how these “anomalies”  are all over the place and could be located easily. It literally took me 2 minutes  to find that example, and have not taken time to search for more.  

The facts are that these “anomalies”  are not anomalies  at all. Rather they are intentional crafting of the text intended to communicate not just pronunciation, but commentary as well. 

There are even versions of the MT on sefaria.org that have the cantillation marks without the vowels, and I have not even tested this table to evaluate how it performs in that arena. There are several areas that need to be tested.  

In previous tables, I have posted links to a mountain of articles which demonstrate that these “anomalies”  are intentional, but it appears that they were  not taken into consideration. 

Indeed, there are times where anomalies are accidental from the publisher, but often they are because there are times the publisher doesn’t understand and are just trying to reproduce digitally what is there in print or hand writing.  

These comments thus far are only about most popular classical literature, we have not even gotten into some other pieces of literature with cantillation marks read in other liturgic 
settings. 


I am very happy and extremely thankful that these guys are working on this table. This table should move forward, but this table should in no way be considered a finished project. I would like to see this table used by native Hebrew speakers as well. In order to do that, we have to work to include all issues represented in the texts of this genre. 

Thank you to all of you for your work.

EricJHarvey commented:
@bertfrees Here is a more detailed description of the problem. Hopefully this is enough to make sense.

There are three categories of character to keep in mind here: consonants, vowels, and accents. I have made user-defined categories for these:
Consnants are defined as %hebLetter
Vowels are %vowel
Accents are %extCant. These are two-cell dot patterns with the first cell being @46.
Here is lou_trace for a correct translation.
מִזְמֹ֥ור
⠍⠊⠵⠄⠍⠨⠍⠪⠗
1.	correct	""	?
2.	letter	מ	134
3.	sign	ִ	24A
4.	letter	ז	1356
5.	sign	ְ	3
6.	letter	מ	134
7.	sign	ֹ	135A
8.	sign	֥	46-134
9.	letter	ו	2456
10.	letter	ר	1235
11.	pass3	@135A[@46$sldpUumSDwxyz]@2456	*@246
12.	pass4	@24A	@24

Steps 7 through 11 are the relevant section, with step  11 being the main issue. Right now, it applies whether the character following the waw (@2456) is a vowel or a consonant. Technically, it should only apply to cases where the next character is a consonant. (%hebLetter) and should exclude cases where the next letter is any other type of character. 
Is there a way to make that specification in the test part and still reproduce both the specific accent and the specific consonant?

Here is a test I made up that fails. This is hypothetical because the accent is never placed in this location in the Hebrew Bible.

יְהֹ֣וָה
⠚⠄⠓⠨⠥⠪⠣⠓
1.	letter	י	245
2.	sign	ְ	3
3.	letter	ה	125
4.	sign	ֹ	135A
5.	sign	֣	46-136
6.	letter	ו	2456
7.	sign	ָ	126
8.	letter	ה	125
9.	pass3	@135A[@46$sldpUumSDwxyz]@2456	*@246

Ideally, this should translate as
⠚⠄⠓⠕⠨⠥⠺⠣⠓
That is, the way it would be without the rule in step 9 applying.

As a side note, @LeonarddeR suggests that I use the correct pass for this, but I am reticent to use correct for character reordering, since the way it consumes characters can interfere with some of the other operations in that pass.

Thanks for your help!

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -22,476 +22,94 @@
 
-include he-common-consonants.uti
+# Define vowel class (also includes Etnahta and zaqef qatan)
+attribute vowel /x0591/x0594\x05b0\x05b1\x05b2\x05b3\x05b4\x05b5\x05b6\x05b7\x05b8\x05b9\x05ba\x05bb\x05c7
```
I just discovered that this contains slashes where I think they should be backslashes

```suggestion
attribute vowel \x0591\x0594\x05b0\x05b1\x05b2\x05b3\x05b4\x05b5\x05b6\x05b7\x05b8\x05b9\x05ba\x05bb\x05c7
```

EricJHarvey commented:
@bertfrees Perhaps it would be easier to compare these two rules. One covers the situation wher where there is no accent on the syllable, and the other covers the situation where there is an accent.
noback pass3 [@135a-2456]%hebLetter @246
noback pass3 @135a[@46$a]@2456 *@246

Is there a way to specify that the second rule should only apply when the waw is followd by a %hebLetter? In my understanding, the use of the * in the action part makes this impossible.

bertfrees commented on tests/braille-specs/hbo.yaml:
```diff
@@ -32,63 +172,35 @@ tests:
   - ["וּבְכָל־נַפְשְׁךָ֖", "⠬⠧⠄⠡⠣⠇⠤⠝⠉⠋⠄⠩⠄⠡⠣⠨⠞"]
   - ["וּבְכָל־מְאֹדֶֽךָ׃", "⠬⠧⠄⠡⠣⠇⠤⠍⠄⠁⠕⠙⠑⠈⠡⠣⠲"]
-
-  # The following tests show problems with the Hebrew writing
-  # system. The problem with Hebrew is that it is written with a 4
-  # stanza Unicode value. Hebrew is written with the 4 following
-  # stanzas. 1 consonant, 2 dagesh which indicates pronunciation for
-  # the consonant, 3 a vowel, and 4 a cantillation mark. Adding to
-  # this complexity, sometimes these stanzas are presented visually in
-  # the correct order for sighted readers, but in Unicode, this order
-  # could be arranged in a couple of different ways. For example: 1
-  # consonant, 2 dagesh, 3 vowel. Another way that is just as common:
-  # 1 consonant, 2 vowel, 3 dagesh. This happens for several
-  # reasons. Sometimes the arrangement is accidental, and other times
-  # it is intentional. Sometimes this happens as a result of
-  # converting one type of a document into another, such as a Google
-  # document being converted into a Word document.
-  #
-  # For this reason, this test document presents portions of text
-  # which evaluate all possibilities of the above mentioned order to
-  # ensure correct and accurate braille, in every circumstance.
-  # Further, the current Hebrew table does not reflect accurate
-  # braille for the orders of the unicode stanza mentioned above.
-
-  # These tests attempts to show how the current Hebrew table does
-  # accurately produce the desired braille for the consanant, vowel
-  # dagesh.
-  - ["בָּרָא", "⠃⠣⠗⠣⠁"]
-  - ["אֱלֹהִים", "⠁⠢⠇⠕⠓⠔⠍"]
-  - ["הָאָרֶץ.", "⠓⠣⠁⠣⠗⠑⠮⠲"]
-  - ["הָיְתָה", "⠓⠣⠚⠄⠹⠣⠓"]
-  - ["וְהָאָרֶץ", "⠺⠄⠓⠣⠁⠣⠗⠑⠮"]
-  - ["הַמָּיִם.", "⠓⠉⠐⠍⠣⠚⠊⠍⠲"]
-
-  # These tests are derived from the BHS and include notes from the
-  # Masorah, and the cantillation marks. These tests also include all
-  # of the Unicode characters and their unique orders, as mentioned
-  # above. The current Hebrew table does not accurately produce the
-  # desired braille for the consanant, vowel, dagesh characters.
```
@EricJHarvey @yeatersink Double checking: are the issues that Matt raised in this PR related to the comments that were removed here? "Various character orders", "stanzas", "cantillation marks that are post or pre positive", ... Unfortunately I am not able to follow any of this. So I rely completely on you guys to agree and tell me whether I can go ahead with the PR or not. Does everyone agree that this PR is an improvement and that no useful stuff (including tests and comments) is deleted?

Of course the table does not need to be perfect yet, nor do we need to agree yet on how exactly we are going to improve it in future versions.

bertfrees commented:
@EricJHarvey Your last commit made the tests fail.

EricJHarvey commented:
Yes, this is the general issue that causes problems in specific cases. At this point, all of the surfaced errors included in the previous yaml have been fixed, and the new issue that Matt raised has also been fixed and tests added. From my perspective, this PR is ready to merge, and we will address any further issues we find when using the tables out in the real world. Thanks!Eric J. Harvey, Ph.D.He/hisAffiliated ResearcherCenter for Spatial and Textual AnalysisStanford Universitywww.blindscholar.comOn Jun 5, 2025, at 7:53 AM, Bert Frees ***@***.***> wrote:﻿
@bertfrees commented on this pull request.



In tests/braille-specs/hbo.yaml:
> -  # The following tests show problems with the Hebrew writing
-  # system. The problem with Hebrew is that it is written with a 4
-  # stanza Unicode value. Hebrew is written with the 4 following
-  # stanzas. 1 consonant, 2 dagesh which indicates pronunciation for
-  # the consonant, 3 a vowel, and 4 a cantillation mark. Adding to
-  # this complexity, sometimes these stanzas are presented visually in
-  # the correct order for sighted readers, but in Unicode, this order
-  # could be arranged in a couple of different ways. For example: 1
-  # consonant, 2 dagesh, 3 vowel. Another way that is just as common:
-  # 1 consonant, 2 vowel, 3 dagesh. This happens for several
-  # reasons. Sometimes the arrangement is accidental, and other times
-  # it is intentional. Sometimes this happens as a result of
-  # converting one type of a document into another, such as a Google
-  # document being converted into a Word document.
-  #
-  # For this reason, this test document presents portions of text
-  # which evaluate all possibilities of the above mentioned order to
-  # ensure correct and accurate braille, in every circumstance.
-  # Further, the current Hebrew table does not reflect accurate
-  # braille for the orders of the unicode stanza mentioned above.
-
-  # These tests attempts to show how the current Hebrew table does
-  # accurately produce the desired braille for the consanant, vowel
-  # dagesh.
-  - ["בָּרָא", "⠃⠣⠗⠣⠁"]
-  - ["אֱלֹהִים", "⠁⠢⠇⠕⠓⠔⠍"]
-  - ["הָאָרֶץ.", "⠓⠣⠁⠣⠗⠑⠮⠲"]
-  - ["הָיְתָה", "⠓⠣⠚⠄⠹⠣⠓"]
-  - ["וְהָאָרֶץ", "⠺⠄⠓⠣⠁⠣⠗⠑⠮"]
-  - ["הַמָּיִם.", "⠓⠉⠐⠍⠣⠚⠊⠍⠲"]
-
-  # These tests are derived from the BHS and include notes from the
-  # Masorah, and the cantillation marks. These tests also include all
-  # of the Unicode characters and their unique orders, as mentioned
-  # above. The current Hebrew table does not accurately produce the
-  # desired braille for the consanant, vowel, dagesh characters.

@EricJHarvey @yeatersink Double checking: are the issues that Matt raised in this PR related to the comments that were removed here? "Various character orders", "stanzas", "cantillation marks that are post or pre positive", ... Unfortunately I am not able to follow any of this. So I rely completely on you guys to agree and tell me whether I can go ahead with the PR or not. Does everyone agree that this PR is an improvement and that no useful stuff (including tests and comments) is deleted?
Of course the table does not need to be perfect yet, nor do we need to agree yet on how exactly we are going to improve it in future versions.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
